File: src/main/java/com/zerozero/core/domain/BaseEntity.java
package com.zerozero.core.domain;

import com.fasterxml.uuid.Generators;
import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.SuperBuilder;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;
import java.util.UUID;

@Getter
@MappedSuperclass
@EntityListeners(value = AuditingEntityListener.class)
@SuperBuilder
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor(access = AccessLevel.PROTECTED)
public class BaseEntity {

    @CreatedDate
    private LocalDateTime createdAt;

    @LastModifiedDate
    private LocalDateTime updatedAt;

    @Id
    @Column(columnDefinition = "BINARY(16)")
    private UUID id;

    @Builder.Default
    private Boolean deleted = false;

    @PrePersist
    public void initSequentialUUID() {
        final String UUID_DELIMITER = "-";
        UUID uuid = Generators.timeBasedGenerator().generate();
        String[] uuidSplitArray = uuid.toString().split(UUID_DELIMITER);
        String sequentialUUIDString = uuidSplitArray[2] + uuidSplitArray[1] + uuidSplitArray[0] + uuidSplitArray[3] + uuidSplitArray[4];
        this.id = UUID.fromString(
                new StringBuilder(sequentialUUIDString)
                        .insert(8, UUID_DELIMITER)
                        .insert(13, UUID_DELIMITER)
                        .insert(18, UUID_DELIMITER)
                        .insert(23, UUID_DELIMITER)
                        .toString()
        );
    }
}

File: src/main/java/com/zerozero/core/domain/BaseEntity.kt
package com.zerozero.core.domain

import com.github.f4b6a3.uuid.UuidCreator
import jakarta.persistence.Column
import jakarta.persistence.EntityListeners
import jakarta.persistence.Id
import jakarta.persistence.MappedSuperclass
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.time.LocalDateTime
import java.util.*

@MappedSuperclass
@EntityListeners(AuditingEntityListener::class)
open class BaseEntity {

    @Id
    @Column(columnDefinition = "BINARY(16)")
    val id: UUID = UuidCreator.getTimeOrderedEpoch()

    @CreatedDate
    @Column(name = "created_at", updatable = false)
    var createdAt: LocalDateTime? = null

    @LastModifiedDate
    @Column(name = "updated_at")
    var updatedAt: LocalDateTime? = null

    @Column(name = "deleted", nullable = false)
    var deleted: Boolean = false
}
